Enhances I18NCreator for creating Message interfaces from properties

tsschmidt opened the pull request:
Currently I18NCreator creates method signatures that only use String arguments regardless of any MessageFormat that may be present for an argument. This patch attempts to create more accurate interface signatures based on the MessageFormat and also includes some GWT specific formats such as {0,list},{0,localdatetime,..} and Plurals.

This patch also proposes to add the convention {0,safehtml} to mark that an argument to be of type SafeHtml.  If an Argument is found to be of type SafeHtml the return type of the method will be set to SafeHtml.

Fixes #9049

tsschmidt committed:
Enhances I18NCreator for creating Message interfaces from properties

Currently I18NCreator creates method signatures that only use String
arguments regardless of any MessageFormat that may be present for
an argument. This patch attempts to create more accurate interface
signatures based on the MessageFormat and also includes some GWT
specific formats such as {0,list},{0,localdatetime,..} and Plurals.

This patch also proposes to add the convention {0,safehtml} to mark that
an argument to be of type SafeHtml.  If an Argument is found to be of
type SafeHtml the return type of the method will be set to SafeHtml.

niloc132 commented on user/src/com/google/gwt/i18n/rebind/AbstractLocalizableInterfaceCreator.java:
```diff
@@ -281,12 +275,12 @@ private void genMethodDecl(String type, String defaultValue, String key) {
     composer.print(");\n");
   }
-
+  
```
Extra spaces still here I'm afraid
```suggestion

```

niloc132 commented on user/src/com/google/gwt/i18n/rebind/MessagesInterfaceCreator.java:
```diff
@@ -100,4 +136,96 @@ protected String javaDocComment(String path) {
+        composer.println(",");
+      }
+      composer.print("\"" + key.substring(key.indexOf('[') + 1, key.length() - 1) + "\", ");
+      composer.print("\"" + plurals.get(key) + "\"");
```
Probably should use makeJavaString here - in theory the keys never have strings in them, but I _think_ it is technically allowed for properties files to have spaces in their keys if escaped?

niloc132 reviewed (commented):
This appears to be in the style of other existing Creator types, so I can't fault it there, but I would generally like the test to be a bit more than 'it can run and doesn't fail in obvious ways' - at a glance, it doesn't even check if the generated sources compile?

A few things to consider below - thank you for getting this patch going again, I'd be inclined to merge it without further _additional_ delay. If you'd like a more rigorous review, I'd be happy to.

tsschmidt committed:
Merge branch 'gwtproject:main' into main

tsschmidt committed:
Enhances I18NCreator for creating Message interfaces from properties

Currently I18NCreator creates method signatures that only use String
arguments regardless of any MessageFormat that may be present for
an argument. This patch attempts to create more accurate interface
signatures based on the MessageFormat and also includes some GWT
specific formats such as {0,list},{0,localdatetime,..} and Plurals.

This patch also proposes to add the convention {0,safehtml} to mark that
an argument to be of type SafeHtml.  If an Argument is found to be of
type SafeHtml the return type of the method will be set to SafeHtml.

niloc132 commented on user/src/com/google/gwt/i18n/rebind/AbstractLocalizableInterfaceCreator.java:
```diff
@@ -250,11 +240,15 @@ void generateFromPropertiesFile() throws IOException {
               + resourceFile
               + "' cannot be used to generate message classes, as it has no key/value pairs defined.");
```
this one is outside your diff, so you can ignore it

tsschmidt committed:
Enhances I18NCreator for creating Message interfaces from properties

Currently I18NCreator creates method signatures that only use String
arguments regardless of any MessageFormat that may be present for
an argument. This patch attempts to create more accurate interface
signatures based on the MessageFormat and also includes some GWT
specific formats such as {0,list},{0,localdatetime,..} and Plurals.

This patch also proposes to add the convention {0,safehtml} to mark that
an argument to be of type SafeHtml.  If an Argument is found to be of
type SafeHtml the return type of the method will be set to SafeHtml.

niloc132 reviewed (approved):
Tests are passing - approving, will leave it up a day or so for other comment.
